### PR TITLE
Some error cleanup steps in the decomposed FS

### PR DESCRIPTION
**Pull request by dragotin**
Some improvements to error handling that tries to avoid changes that were made before (ie. new directories) in case of a later error.

### CODE REVIEW

File: pkg/storage/utils/decomposedfs/decomposedfs.go [hunk 1]
```diff
@@ -363,39 +378,62 @@ func (fs *Decomposedfs) CreateReference(ctx context.Context, p string, targetURI
 	}
 
 	// create Shares folder if it does not exist
-	var n *node.Node
-	if n, err = fs.lu.NodeFromResource(ctx, &provider.Reference{Path: fs.o.ShareFolder}); err != nil {
+	var parentNode *node.Node
+	var parentCreated, childCreated bool // defaults to false
```
Comment (C0rby) [inline pkg/storage/utils/decomposedfs/decomposedfs.go:382]: Cool! Better naming. :+1:

File: pkg/storage/utils/decomposedfs/decomposedfs.go [hunk 2]
```diff
@@ -363,39 +378,62 @@ func (fs *Decomposedfs) CreateReference(ctx context.Context, p string, targetURI
...
+	}
+
+	var childNode *node.Node
+	// clean up directories created here on error
+	defer func() {
+		if err != nil {
+			// do not catch the error to not shadow the original error
+			if childCreated && childNode != nil {
+				if tmpErr := fs.tp.Delete(ctx, childNode); tmpErr != nil {
+					appctx.GetLogger(ctx).Error().Err(tmpErr).Msg("Can not clean up child node after error")
```
Comment (C0rby) [inline pkg/storage/utils/decomposedfs/decomposedfs.go:402]:
```suggestion
					appctx.GetLogger(ctx).Error().Err(tmpErr).Str("node_id", childNode.ID).Msg("Can not clean up child node after error")
```

File: pkg/storage/utils/decomposedfs/decomposedfs.go [hunk 3]
```diff
@@ -363,39 +378,62 @@ func (fs *Decomposedfs) CreateReference(ctx context.Context, p string, targetURI
...
+				}
+			}
+			if parentCreated && parentNode != nil {
+				if tmpErr := fs.tp.Delete(ctx, parentNode); tmpErr != nil {
+					appctx.GetLogger(ctx).Error().Err(tmpErr).Msg("Can not clean up parent node after error")
```
Comment (C0rby) [inline pkg/storage/utils/decomposedfs/decomposedfs.go:407]:
```suggestion
					appctx.GetLogger(ctx).Error().Err(tmpErr).Str("node_id", parentNode.ID).Msg("Can not clean up parent node after error")
```

**Review by C0rby — APPROVED** [review]
Only small remarks.

**Review by butonic — APPROVED** [review]

File: pkg/storage/utils/decomposedfs/spaces.go [hunk 4]
```diff
@@ -505,13 +505,15 @@ func (fs *Decomposedfs) createStorageSpace(ctx context.Context, spaceType, space
...
+			// FIXME: is it ok to wipe this err if the symlink already exists?
+			err = nil
 		} else {
 			// TODO how should we handle error cases here?
 			appctx.GetLogger(ctx).Error().Err(err).Str("space", spaceID).Str("spacetype", spaceType).Msg("could not create symlink")
 		}
 	}
 
-	return nil
+	return err
```
[the remark this answers is not in the document]
Comment (dragotin) [reply pkg/storage/utils/decomposedfs/spaces.go:516]: No. In this case, it existed from before, and we only should remove stuff that we did in the function here before.

**Pull request merged**